pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/internal/ImportOnDemandSymbolTable.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.symbols.table.internal;
+
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.lang.java.ast.ASTImportDeclaration;
+import net.sourceforge.pmd.lang.java.symbols.JClassSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JFieldSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JMethodSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JTypeDeclSymbol;
+import net.sourceforge.pmd.lang.java.symbols.table.JSymbolTable;
+import net.sourceforge.pmd.lang.java.symbols.table.ResolveResult;
+import net.sourceforge.pmd.lang.java.symbols.table.internal.ResolveResultImpl.ClassResolveResult;
+
+
+/**
+ * Scope for imports on demand. Imports-on-demand never shadow anything, so this scope, if it exists,
+ * is the top-level non-empty scope. All scope stacks have {@link EmptySymbolTable} as bottom though, for
+ * implementation simplicity.
+ *
+ * @since 7.0.0
+ */
+final class ImportOnDemandSymbolTable extends AbstractImportSymbolTable {
+
+    /** Stores the names of packages and types for which all their types are imported. */
+    private final Map<String, ASTImportDeclaration> importedPackagesAndTypes = new HashMap<>();
+
+
+    /**
+     * @param importsOnDemand List of import-on-demand statements, mustn't be single imports!
+     */
+    ImportOnDemandSymbolTable(JSymbolTable parent, SymbolTableHelper helper, List<ASTImportDeclaration> importsOnDemand) {
+        super(parent, helper);
+
+        for (ASTImportDeclaration anImport : importsOnDemand) {
+            assert anImport.isImportOnDemand() : "Expected import on demand: " + anImport;
```

Comment:
Maybe throw a IllegalArgumentException like in SingleImportSymbolTable?

Reply:
Maybe the other one should be an assert instead... In fact maybe we need to rationalize some guidelines about this? I tend to use `assert` statements for simple sanity checks, that should fail during testing if ever. For example here, the dataflow leading to this constructor is simple, entirely known, and doesn't allow anything else to happen. OTOH I use IllegalArgumentException for visible API, where a method is expected by a caller to check its arguments (and we can't know the whole dataflow leading to it, because the method is somehow public). Meaning, an assert tests the sanity of the upstream data flow, whereas an IllegalArgumentException means it's expected that someday someone will throw something illegal at it. Not sure this makes complete sense..